### PR TITLE
Issue 510d: add DataspacePool and DataspaceHolder to speed-up h5cpp

### PR DESCRIPTION
**Pull request by jkotan**
It resolves (d) of #510 by adding `DataspacePool` and `DataspaceHolder` to speed-up the `Dataset.write/read`. In this way we avoid creating/destroying `Dataspace` objects which is quite time consuming. 
Usually user dataspaces are of similar type/size for one `Dataset`, so the `Dataspace` pool is added as a member of `Dataset` class.
After applying the PR with  #511, #512, #513 the original test example
```c++
    std::vector<long int> vc(1);
    for(size_t i = 0; i < 1000000; i++) {
      vc[0] = static_cast<long int>(std::rand());
      dataset.resize({i + 1});
      selection.offset(0, i);
      dataset.write(vc, selection);
    }
```
speed-up ~ 4 times .

### CODE REVIEW

File: src/h5cpp/dataspace/dataspace.cpp [hunk 1]
```diff
@@ -51,19 +51,27 @@ Dataspace::Dataspace(ObjectHandle &&handle)
...
+    hid_t ret = H5Scopy(static_cast<hid_t>(space.handle_));
+    if (0 > ret) {
     error::Singleton::instance().throw_with_stack("could not copy Dataspace");
+    }
+    handle_ = ObjectHandle(ret);
   }
-  handle_ = ObjectHandle(ret);
+  else
+    handle_ = ObjectHandle();
   return *this;
```
Comment (yuelongyu) [inline src/h5cpp/dataspace/dataspace.cpp:75]: another swap method for this, like the one in DataType::swap?

File: src/h5cpp/node/dataset.cpp [hunk 2]
```diff
@@ -189,10 +189,10 @@ void Dataset::refresh() const
 
 #if H5_VERSION_GE(1,10,2)
 
-long long unsigned int Dataset::chunk_storage_size(
-			 std::vector<long long unsigned int> offset) const
+unsigned long long Dataset::chunk_storage_size(
```
Comment (yuelongyu) [inline src/h5cpp/node/dataset.cpp:192]: what about using uint64_t instead of unsigned long long

File: src/h5cpp/dataspace/type_trait.hpp [hunk 3]
```diff
@@ -55,6 +58,11 @@ class TypeTrait {
     return Scalar();
   }
 
+  const static DataspaceType & get(const T &, DataspacePool &) {
+    const static DataspaceType & cref_ = Scalar();
+    return cref_;
+  }
```
Comment (yuelongyu) [inline src/h5cpp/dataspace/type_trait.hpp:64]: DataspacePool is not used in the function, why it's here?

Comment (jkotan) [reply src/h5cpp/dataspace/type_trait.hpp:64]: It is becuase to have the common interface with other `get()` methods of the factory. The `pool` is used for `Simple` datatypes e.g. with `vector<T>` Type.

File: src/h5cpp/node/dataset.cpp (hunk 2 above)
Comment (jkotan) [reply src/h5cpp/node/dataset.cpp:192]: In the API of libhdf5 the offset is defined with `hsize_t`. I will try to switch to it

**Review by yuelongyu — APPROVED** [review]

**Pull request merged**